astroid/nodes/scoped_nodes/scoped_nodes.py

```diff
@@ -3254,3 +3267,17 @@ def frame(self: T, *, future: Literal[None, True] = None) -> T:
         :returns: The node itself.
         """
         return self
+
+    @decorators_mod.cachedproperty
+    def tolineno(self) -> Optional[int]:
+        """The last line that this node appears on in the source code."""
+        if self.end_lineno is not None:
+            return self.end_lineno
+        last_child = self.last_child()
+        if last_child is None:
```

Comment:
Same question for `ClassDef`, the possible fields `("decorators", "bases", "keywords", "body")`.

Reply:
https://github.com/PyCQA/astroid/blob/56f5f055e4847e9fc2b74162ccad8f2a11db3bfc/astroid/nodes/node_ng.py#L445-L457

This is just copied from the original implementation. Only thing that changes here is the `if self.doc` line.

Reply:
What about this?
```py
class A(B, var=42):
    """Hello World"""
```

Reply:
Again, there are ways to work around it, although slightly more complicated once here.
We need:
* Some way to filter `decorators`
* Make sure `last_child` is not a `body` node
* And if `last_child` is not None but we don't have a `doc`, we probably shouldn't fall back to `self.fromlineno`.